### PR TITLE
Strip HTML from gallery image alt text

### LINKED ISSUE #26549: Gallery block: Strip HTML from caption before using as alt attribute.
**Describe the bug**
Per #26082, in the gallery block, for an image without an `alt` text specified, we're using the image's caption as a fallback. However, the caption can contain HTML (e.g. a link, i.e. an `<a />` tag) that we should strip from it when using it as `alt` text.

**To reproduce**
1. Insert a gallery block.
2. Add a few images.
3. Add a caption to one of the images.
4. Highlight part of the caption and add a link (e.g. to wordpress.org).
5. Publish the post, and view it one the frontend.
6. Inspect the page source.
7. Verify that the image's `alt` attribute includes the HTML (`<a />` element).

**Expected behavior**
HTML should be removed from the caption when used as `alt` attribute.

**Screenshots**
N/A

**Editor version (please complete the following information):**
- WordPress version: 5.5
- Does the website has Gutenberg plugin installed, or is it using the block editor that comes by default? Gutenberg plugin v9.2.2 installed.
- If the Gutenberg plugin is installed, which version is it? [e.g., 7.6]

Brought up [here](https://github.com/WordPress/gutenberg/pull/26082#issuecomment-718078696). When used with other plugins, there can be issued with incorrectly mangled HTML in attributes, see e.g. https://github.com/Automattic/wp-calypso/issues/46813.

cc/ @ZebulanStanphill

### PR DESCRIPTION
**Pull request by WunderBart**
Fixes https://github.com/WordPress/gutenberg/issues/26549

## Description
When a gallery image's `alt` attribute value is empty, we use what's in the `caption` prop to fill it out. The issue is that `caption` can contain links, whereas the `alt` attribute should not contain any HTML. This PR makes sure the `alt` attribute when created from `caption`, receives a value stripped out of HTML.

## How has this been tested?
Manually:
1. Add the Image Gallery block.
2. Add some images.
3. Fill the caption with some text and a link.
4. Preview the site and inspect the image element.
5. See that the `alt` attribute doesn't contain any HTML.

To run unit tests covering the change:

```
npm run test-unit -- test/integration/full-content/full-content.test.js
```

## Types of changes
Bugfix

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style.

## TODO
- [ ] Figure out a way to access 'implicit' attributes (that aren't stored in the comment, but inferred through a query) in `migrate`
- [ ] Migration: Unescape HTML in `alt` attribute.

### CODE REVIEW

File: packages/block-library/src/gallery/gallery-image.js
Comment (ockham) [inline packages/block-library/src/gallery/gallery-image.js:287]: We don't want to necessarily override the `alt` which is a different attribute. I.e. a user can provide an `alt` and a `caption` (or no `caption`), in which case we'll use the alt and won't even look at the `caption`.

So we should strip the HTML from `caption` at save time, here: https://github.com/WordPress/gutenberg/pull/26082/files#diff-4283002b00fa661721fcb36f9eb609725082e259139407334ab318dd46746167R46

Comment (WunderBart) [reply packages/block-library/src/gallery/gallery-image.js]: Updated in https://github.com/WordPress/gutenberg/pull/26574/commits/898b5f89cf2e81d89d815b3ac59d98e575f01300.

**Comment by ockham** [comment]
The current error 

```
    Expected mock function not to be called but it was called with:
    ["Block successfully updated for `%s` (%o).·
    New content generated by `save` function:·
    [...]
```

seems to indicate that a block migration is run, even though it shouldn't be run :thinking: 

BTW how did you create the fixtures? https://github.com/WordPress/gutenberg/blob/710373b254fbcd15d524afdeb31da0d93c4defd4/packages/e2e-tests/fixtures/blocks/README.md#creating-fixtures says you only need to manually create `core__gallery__image-with-caption.html`, and can then auto-generate the other fixtures through

```
npm run fixtures:regenerate test/integration/full-content/full-content.test.js
```

I had a look at `core__gallery__image-with-caption.html`. I _think_ that the `alt` img attribute cannot be empty there, since we're assuming that this block was saved using the current (as of this PR) block code -- which would set the `alt` img attribute to `caption`'s (HTML-stripped) value.

**Comment by ockham** [comment]
> I had a look at `core__gallery__image-with-caption.html`. I _think_ that the `alt` img attribute cannot be empty there, since we're assuming that this block was saved using the current (as of this PR) block code -- which would set the `alt` img attribute to `caption`'s (HTML-stripped) value.

Yeah, setting those (and regenerating the other fixtures) seems to do the trick :+1:

**Comment by WunderBart** [comment]
> > I had a look at core__gallery__image-with-caption.html. I think that the alt img attribute cannot be empty there, since we're assuming that this block was saved using the current (as of this PR) block code -- which would set the alt img attribute to caption's (HTML-stripped) value.
>
> Yeah, setting those (and regenerating the other fixtures) seems to do the trick 👍

I did use the `npm` command to create the remaining files. Missed the fact that the `alt` attribute should be filled with the output value there, though! 🙌. Thanks for fixing the tures, Bernie! 😄

File: packages/block-library/src/gallery/deprecated.js
Comment (ockham) [inline packages/block-library/src/gallery/deprecated.js:117]: We need to return the entire image object (with the modified/unmodified `alt`) here, no? :smile:

Comment (WunderBart) [reply packages/block-library/src/gallery/deprecated.js:117]: We absolutely, 100% do! 😅

Comment (WunderBart) [reply packages/block-library/src/gallery/deprecated.js]: Fixed in https://github.com/WordPress/gutenberg/pull/26574/commits/82c2a6f0aecf54d9ebd75343d0c25cab555b102b

**Comment by ockham** [comment]
Writing a deprecation here is currently **blocked** by the fact that apparently, 'implicit' attributes (i.e. the ones that aren't stored as part of the block comment, but inferred through a queue) -- in this case: the `images` array -- aren't accessible to `isEligible` and `migrate`.

I think that's a feature that should be added to deprecations, although it might be a bit tricky to get right, since serializing the updated attribute back to HTML would involve calling the right `save` method (not the current deprecation's). We can probably assume it's the 'next' one in the stack of deprecations (including the up-to-date `save` of the block).

cc/ @youknowriad @mcsf @gziolo

**Comment by youknowriad** [comment]
> Writing a deprecation here is currently blocked by the fact that apparently, 'implicit' attributes (i.e. the ones that aren't stored as part of the block comment, but inferred through a queue) -- in this case: the images array -- aren't accessible to isEligible and migrate.

How did you arrive to this conclusion. That's not true I think or unexpected.

File: packages/block-library/src/gallery/deprecated.js
Comment (youknowriad) [inline packages/block-library/src/gallery/deprecated.js:102]: Why do you need isElligible here? I believe this particular deprecation doesn't really need it?

File: packages/block-library/src/gallery/deprecated.js
Comment (youknowriad) [inline packages/block-library/src/gallery/deprecated.js:126]: Also, I'm not sure the migrate function is needed here because if I understand the PR properly, the attribute is not really rewritten, the same attribute will produce another output, that's it.

File: packages/block-library/src/gallery/deprecated.js
Comment (ockham) [reply packages/block-library/src/gallery/deprecated.js:102]: > > Writing a deprecation here is currently blocked by the fact that apparently, 'implicit' attributes (i.e. the ones that aren't stored as part of the block comment, but inferred through a queue) -- in this case: the images array -- aren't accessible to isEligible and migrate.
> 
> How did you arrive to this conclusion. That's not true I think or unexpected.

I put a `console.log` into both functions to output their first two arguments, and the first one only contained the attributes that were listed in the block comment.

```diff
diff --git a/packages/block-library/src/gallery/deprecated.js b/packages/block-library/src/gallery/deprecated.js
index 8542eb5bfc..b0ed9d7ea1 100644
--- a/packages/block-library/src/gallery/deprecated.js
+++ b/packages/block-library/src/gallery/deprecated.js
@@ -99,7 +99,9 @@ const deprecated = [
                        anchor: true,
                        align: true,
                },
-               isEligible( { images } ) {
+               isEligible( props, blcks ) {
+                       console.log( 'isElig', props, blcks );
+                       const { images } = props;
                        return (
                                images &&
                                images.some(
@@ -107,7 +109,8 @@ const deprecated = [
                                )
                        );
                },
-               migrate( attributes ) {
+               migrate( attributes, blcks ) {
+                       console.log( 'migrate', attributes, blcks );
                        return {
                                ...attributes,
                                images: attributes.images.map( ( image ) => {
```

image

**Comment by ockham** [comment]
> > Writing a deprecation here is currently blocked by the fact that apparently, 'implicit' attributes (i.e. the ones that aren't stored as part of the block comment, but inferred through a queue) -- in this case: the images array -- aren't accessible to isEligible and migrate.
> 
> How did you arrive to this conclusion. That's not true I think or unexpected.

I put a `console.log` into both functions to output their first two arguments, and the first one only contained the attributes that were listed in the block comment.

```diff
diff --git a/packages/block-library/src/gallery/deprecated.js b/packages/block-library/src/gallery/deprecated.js
index 8542eb5bfc..b0ed9d7ea1 100644
--- a/packages/block-library/src/gallery/deprecated.js
+++ b/packages/block-library/src/gallery/deprecated.js
@@ -99,7 +99,9 @@ const deprecated = [
                        anchor: true,
                        align: true,
                },
-               isEligible( { images } ) {
+               isEligible( props, blcks ) {
+                       console.log( 'isElig', props, blcks );
+                       const { images } = props;
                        return (
                                images &&
                                images.some(
@@ -107,7 +109,8 @@ const deprecated = [
                                )
                        );
                },
-               migrate( attributes ) {
+               migrate( attributes, blcks ) {
+                       console.log( 'migrate', attributes, blcks );
                        return {
                                ...attributes,
                                images: attributes.images.map( ( image ) => {
```

image

File: packages/block-library/src/gallery/deprecated.js
Comment (ockham) [reply packages/block-library/src/gallery/deprecated.js:126]: Well the `alt` attribute -- which is implicit -- is being changed; so yeah, that attribute is part of the output (and not an explicit attribute in the block comment). Are you saying that a change like that doesn't need a migration? :thinking: If we don't add one, it means the block will load _with_ HTML in the `alt` attribute -- it will only be removed when saving it again, correct?

File: packages/block-library/src/gallery/deprecated.js
Comment (ockham) [reply packages/block-library/src/gallery/deprecated.js:102]: Because of what you said in your other [comment](https://github.com/WordPress/gutenberg/pull/26574#discussion_r515156056)? I guess if we don't need a `migrate`, we won't need an `isEligible`. But if we need a `migrate`, we'd need a way to know that the `alt` includes HTML, no?

File: packages/block-library/src/gallery/deprecated.js
Comment (youknowriad) [reply packages/block-library/src/gallery/deprecated.js:126]: yes, correct. but now that I think about it, there's something weird in the implementation of the gallery save function regardless.

If the image doesn't have any "alt", it will get assigned an "alt" text after saving because it will be parsed from the fallback. I wonder if we should remove the fallback personally and instead apply the fallback in `edit` function instead.

Comment (ockham) [reply packages/block-library/src/gallery/deprecated.js:126]: > yes, correct. but now that I think about it, there's something weird in the implementation of the gallery save function regardless.
> 
> If the image doesn't have any "alt", it will get assigned an "alt" text after saving because it will be parsed from the fallback. I wonder if we should remove the fallback personally and instead apply the fallback in `edit` function instead.

You mean weird in terms of UX? So right now, after saving and reloading, the `alt`, if otherwise empty, would be prefilled with the `caption`. Moving to `edit` would essentially do the same though, no?

Whatever we do, let's please make sure to strip the HTML, and provide a working migration path. This [issue](https://github.com/WordPress/gutenberg/issues/26549) (when combined with some other HTML mangling/sanitization) is causing [havoc](https://github.com/Automattic/wp-calypso/issues/46813).

Comment (youknowriad) [reply packages/block-library/src/gallery/deprecated.js:126]: I mean the block saving is not idempotent. If you save  the same post twice without changes, you get different attributes.

What if we do this:

 - When adding an image, use the caption as a fallback for the alt property. (and escape it)
 - Remove the fallback in save

Would that work?

Comment (ockham) [reply packages/block-library/src/gallery/deprecated.js:126]: I think that makes sense, yeah 👍🏼

Comment (ockham) [reply packages/block-library/src/gallery/deprecated.js:126]: Okay, that means basically going with what @WunderBart [originally had](https://github.com/WordPress/gutenberg/pull/26574#discussion_r514391807).

I'm still not entirely sure what that means for the blocks out there that now have HTML in their `alt` attributes :thinking: Apparently, we can't use a migration to fix them. So I guess we'd need to take care of that removal in (both?) `edit` and `save`? :thinking:

Comment (ockham) [reply packages/block-library/src/gallery/deprecated.js:126]: > I mean the block saving is not idempotent. If you save the same post twice without changes, you get different attributes.
> 
> What if we do this:
> 
> * When adding an image, use the caption as a fallback for the alt property. (and escape it)
> * Remove the fallback in save
> 
> Would that work?

@WunderBart and I just had a call: it won't work exactly as you described, as the caption can be edited in the editor once the image has been inserted. So we'd basically need to update through the caption's `RichText` component's `onChange`. This probably means adding some additional criteria (only update the `alt` if it's empty, _or_ if `alt === __unstableStripHTML( caption )`).

gallery

cc/ @youknowriad

Comment (youknowriad) [reply packages/block-library/src/gallery/deprecated.js:126]: > This probably means adding some additional criteria (only update the alt if it's empty, or if alt === __unstableStripHTML( previousCaption )).

This makes sense yes.

Comment (WunderBart) [reply packages/block-library/src/gallery/deprecated.js]: Thanks for helping with this, folks! 🙌 

> I'm still not entirely sure what that means for the blocks out there that now have HTML in their alt attributes 🤔 Apparently, we can't use a migration to fix them. So I guess we'd need to take care of that removal in (both?) edit and save?

IIUC, without the migration, we're in a situation in which a user needs to edit the block and save again in order to update the `image.alt` value. If that's true and we're OK with it, it should be sufficient to make an update on `save` only, like this:

https://github.com/WordPress/gutenberg/blob/a85b517c5993696f9dc77b747733a48778990113/packages/block-library/src/gallery/save.js#L44-L61

I tested this implementation for both scenarios - newly created `alt` and one that already existed and contained HTML. What do you think? Did I miss something here?

Comment (WunderBart) [reply packages/block-library/src/gallery/deprecated.js]: Here's how I tested the above:

1. Checkout to master
2. Add gallery block
3. Add one image from the Media Library (make sure its `alt` attr is empty)
4. Add a caption to that image and add some HTML to it (make text bold, add link, etc.)
5. Save draft & reload the page
6. Check that image's source - its `alt` should contain HTML tags
7. Checkout this branch & refresh the page
8. Add another image to the same Gallery (also make sure its `alt` is empty)
9. Like previously, add a caption with some HTML in it
10. Save draft & refresh the page
11. Check source for both images - both `alt` attributes should be correctly stripped out of HTML.

Comment (ockham) [reply packages/block-library/src/gallery/deprecated.js]: I know that @youknowriad was originally opposed to a `save`-based approach because of l[ack of idempotency](https://github.com/WordPress/gutenberg/pull/26574#discussion_r515181556) (and suggested an [`edit`-based approach](https://github.com/WordPress/gutenberg/pull/26574#discussion_r515977791) instead):

> If you save  the same post twice without changes, you get different attributes.

Although I'm not so sure anymore that's true?

- In a gallery block, insert an image without an `alt` text.
- Add a caption.
- Save the post. The caption is used for the `alt` attribute.
- Save again. Now that we already have an `alt` attribute, nothing changes.

@youknowriad Could you weigh in here?

Comment (ockham) [reply packages/block-library/src/gallery/deprecated.js]: I guess some of the problem here is due to the fact that the `alt` attribute, if changed from the media modal, is propagated to the `alt_text` property of the image (whereas the same isn't true if set from the editor during `save`).

Since I would really like the issue with HTML in the `alt` attribute (which is [breaking](https://github.com/Automattic/wp-calypso/issues/46813) a lot of our users' galleries) to be fixed in 9.4, I propose that in case of doubt, we settle for a simple solution that we can all agree on and that we know will fix the issue, without introducing any new regressions; and that we figure out a 'better' solution later. I think that @WunderBart's current, `save`-based solution, fulfills that criterion ([see](https://github.com/WordPress/gutenberg/pull/26574#discussion_r516010096)).

Comment (WunderBart) [reply packages/block-library/src/gallery/deprecated.js]: > I know that @youknowriad was originally opposed to a `save`-based approach because of l[ack of idempotency](https://github.com/WordPress/gutenberg/pull/26574#discussion_r515181556) (and suggested an [`edit`-based approach](https://github.com/WordPress/gutenberg/pull/26574#discussion_r515977791) instead):
> 
> > If you save  the same post twice without changes, you get different attributes.
> 
> Although I'm not so sure anymore that's true?

Ah, good point! It's idempotent for newly created `alt`s only. For `alt`s that already exist and contain `HTML` it will produce a different output this one time (migration-like). Can we live with that, @youknowriad?

File: packages/block-library/src/gallery/save.js
Comment (ockham) [inline packages/block-library/src/gallery/save.js:48]: Let's maybe add a comment why we're _also_ sanitizing `image.alt` (and not just `image.caption`). Something like:

```suggestion
					// image.caption comes from a RichText component, so it can contain
					// HTML that we need to strip. In GB 9.2, image.alt would fall back to
					// image.caption without stripping the HTML, so we also run it through
					// __unstableStripHTML here.
					let alt = image.alt || image.caption;

					if ( alt && alt !== __unstableStripHTML( alt ) ) {
						alt = __unstableStripHTML( alt );
					}
```

(Feel free to change the wording if it's not clear enough.)

File: packages/block-library/src/gallery/deprecated.js
Comment (ockham) [reply packages/block-library/src/gallery/deprecated.js]: I'll add that I believe that we need some logic like this in the `save` anyway, for the purpose of the pseudo-migration for `alt` attributes that were touched with GB 9.2 active, and had HTML inserted. (I don't think we can solve that solely at `edit` level.) And since the change (at `save`) level is sufficient also for newly created galleries, there's nothing left to do at `edit` level.

**Review by ockham — APPROVED** [review]
This LGTM (but we should look into the failing tests -- maybe a rebase is enough).

I'd like to get final approval by @youknowriad since we ended up going with a `save`-based approach; see the conversation starting [here](https://github.com/WordPress/gutenberg/pull/26574#discussion_r516005750) for the rationale.

Thanks a lot for this fix, @WunderBart!

File: packages/block-library/src/gallery/deprecated.js
Comment (youknowriad) [reply packages/block-library/src/gallery/deprecated.js]: > I'll add that I believe that we need some logic like this in the save anyway, for the purpose of the pseudo-migration for alt attributes that were touched with GB 9.2 active, and had HTML inserted.

For me personally, I don't see that as a big issue. I see the conceptual issue as more important: (the conceptual issue being the save function not idempotent creating different block attributes without touching the editor).

How many persons are affected by this issue and for Core, it's still a development plugin.

Comment (youknowriad) [reply packages/block-library/src/gallery/deprecated.js]: I'd also note that in all cases, the user has to open and save the post anyway since migrations don't run automtically.

Comment (WunderBart) [reply packages/block-library/src/gallery/deprecated.js]: > For me personally, I don't see that as a big issue. I see the conceptual issue as more important: (the conceptual issue being the save function not idempotent creating different block attributes without touching the editor).

What if we "mapped" the `alt` prop in `gallery.image.js`'s `compose` instead?

```js
export default compose( [
	withSelect( ( select, ownProps ) => {
		const { getMedia } = select( 'core' );
		const { id } = ownProps;
		const image = id ? getMedia( parseInt( id, 10 ) ) : null;

		if ( image ) {
			let alt = image.alt_text || image.alt || image.caption;

			if ( alt && alt !== __unstableStripHTML( alt ) ) {
				alt = __unstableStripHTML( alt );
			}

			image.alt = alt;
		}

		return {
			image,
		};
	// ...
```

This would handle `alt` fallback but the user would see what they're saving this way. It still requires setting `alt` on caption edit but that's not a problem.

Comment (ockham) [reply packages/block-library/src/gallery/deprecated.js]: > For me personally, I don't see that as a big issue. I see the conceptual issue as more important: (the conceptual issue being the save function not idempotent creating different block attributes without touching the editor).

I'm all for purity of code and all that, but it's the impact of this issues that worries me: Post content actually gets broken, so people will have to revert to an earlier revision (and might lose the changes they were working on). OTOH, the feature that was added by the [PR](https://github.com/WordPress/gutenberg/pull/26082) that caused this is an (arguably minor) a11y improvement. IMO, those aren't strong enough reasons to keep around a feature that has such negative side effects.

> How many persons are affected by this issue and for Core, it's still a development plugin.

Please refer to https://github.com/Automattic/wp-calypso/issues/46813. I'm counting ~15 so far (some of those comments refer to the WP.com forums, where each thread can potentially contain multiple reports. Here's a few:

https://wordpress.com/forums/topic/gallery-block-problem/
https://wordpress.com/forums/topic/galleries-not-showing-up-right/
https://wordpress.com/forums/topic/unreadable-appearance-of-gallery-page-after-editing/
https://wordpress.com/forums/topic/cant-make-changes-to-website-and-drop-down-menus-are-not-working/
https://wordpress.com/forums/topic/galleries-not-showing-up-right/

> I'd also note that in all cases, the user has to open and save the post anyway since migrations don't run automtically.

That's correct, but that's how migrations work, isn't it? This could be addressed by adding additional tooling to e.g. 'fix' the database upon plugin upgrade, but let's not get the lack of a perfect solution in the way of a sufficient one.

Needing to open and save can at least help with the cases where users with GB 9.2 were editing their galleries and because of the issue lost content, revert the post to a previous revision, and would like to re-apply those changes.

---

Anyway, is the compromise to simply remove the `image.caption` fallback for `alt`? I'd be okay with that, for the sake of at least preventing any further breakage into 9.3, even if it means that we won't unbreak existing content.

Comment (ockham) [reply packages/block-library/src/gallery/deprecated.js]: > Anyway, is the compromise to simply remove the `image.caption` fallback for `alt`? I'd be okay with that, for the sake of at least preventing any further breakage into 9.3, even if it means that we won't unbreak existing content.

Filed the alternative PR: #26676

**Comment by gziolo** [comment]
See my comment on the alternative PR proposed by @ockham https://github.com/WordPress/gutenberg/pull/26676#issuecomment-721590464.

**Comment by ockham** [comment]
Obsoleted by #26676.

**Comment by ockham** [comment]
Thanks again for working on this and exploring different options to solve it, @WunderBart!

**Pull request closed without merging**
